// adapt-authoring-lang/lib/module.js
const { AbstractModule, Responder, Utils } = require('adapt-authoring-core');
const fs = require('fs-extra');
const glob = require('glob');
const path = require('path');
const Polyglot = require('node-polyglot');
const util = require('util');

const globPromise = util.promisify(glob);
/**
* Module to handle localisation of language strings
* @extends {AbstractApiModule}
*/
class LangModule extends AbstractModule {
  /** @override */
  constructor(app, pkg) {
    super(app, pkg);
    /**
    * The current locale of the back-end application
    * @type {String}
    */
    this.locale = this.defaultLocale;
    /**
    * The loaded language phrases to be used for translation
    * @type {Object}
    */
    this.phrases = {};
    /**
    * Reference to the Polyglot instance
    * @type {Polyglot}
    */
    this.polyglot = new Polyglot({ phrases: this.phrases, warn: this.logMissingKey.bind(this) });

    this.init();
  }
  /**
  * Tries to determine the environment language from process.env variables.
  * @type {String}
  */
  get defaultLocale() {
    const env = process.env;
    const l = env.LANG || env.LANGUAGE || env.LC_ALL || env.LC_MESSAGES;
    if(!l) {
      l = 'en';
      this.log('warn', `${enErrors.nolang} ${l}`);
      return l;
    }
    return l.slice(0,2);
  }
  /**
  * Returns the languages supported by the application
  * @type {Array<String>}
  */
  get supportedLanguages() {
    return Object.keys(this.phrases);
  }
  /**
  *
  */
  async init() {
    try {
      await this.loadPhrases();
    } catch(e) {
      this.log('error', e);
    }
    this.initLocale();

    return this.setReady();
  }
  /**
  * Switches to the user-configured locale once config has been loaded
  */
  async initLocale() {
    let config;
    try {
      config = await this.app.waitFor('config');
    } catch(e) {
      return this.log('error', e);
    }
    this.locale = this.getConfig('locale');
  }
  * Loads, validates and merges all defined langage phrases
  * @return {Object}
  */
  async loadPhrases() {
    this.phrases = {};
    const deps = [
      { name: this.app.name, rootDir: process.cwd() },
      ...Object.values(this.app.dependencies)
    ];
    return await Promise.all(deps.map(async d => {
      return Object.assign(this.phrases, (await this.loadPhrasesForDir(d.rootDir)));
    }));
  }
  async loadPhrasesForDir(dir) {
    const files = await globPromise(`lang/*.json`, { cwd: dir, absolute: true });
    const strings = {};
    await Promise.all(files.map(async f => {
      const namespace = path.basename(f).replace('.json', '');
      const contents = await fs.readJson(f);
      Object.entries(contents).forEach(([k,v]) => strings[`${namespace}.${k}`] = v);
    }));
    return strings;
  }
  /**
  * Shortcut to log a missing language key
  * @param {String} m The missing key
  */
  logMissingKey(m) {
    const key = m.match(/"(.+)"/)[1];
    this.log('warn', this.phrases[key] ? this.t('error.missingkey', { key }) : m);
  }
  /**
  * Returns translated language string
  * @param {String} key
  * @param {...*} rest
  * @type {Function}
  * @see https://airbnb.io/polyglot.js/#polyglotprototypetkey-interpolationoptions
  */
  t(key, ...rest) {
    this.polyglot.t(`${this.locale}.${key}`, ...rest);
  }
  /**
  * Shorthand function for {@link LangUtility#t} to return translated success language phrases
  * @type {Function}
  * @return {String}
  */
  success(key, ...rest) {
    this.t(`success.${key}`, ...rest);
  }
  /**
  * Shorthand function for {@link LangUtility#t} to return translated error language phrases
  * @type {Function}
  * @return {String}
  */
  error(key, ...rest) {
    this.t(`error.${key}`, ...rest);
  }
}

module.exports = LangModule;
