// adapt-authoring-core/lib/app.js
const DataValidationError = require('./datavalidationerror');
const DependencyLoader = require('./dependencyloader');
const AbstractModule = require('./abstractmodule');
const path = require('path');
const Utils = require('./utils');

let instance;
// /** @ignore */ let initialising = false;
/**
* The main application class
*/
class App extends AbstractModule {
  /**
  * Returns the singleton instance, or initialises it if there isn't one
  * @return {App} The instance
  */
  static get instance() {
    if(!instance) instance = new App();
    return instance;
  }
  /**
  * Create the application instance
  */
  constructor() {
    const pkg = require(path.join(process.cwd(), 'package.json'));
    super(null, { ...pkg, name: 'app', dir: path.join(__dirname, '..') });
    // set app ref here as 'this' doesn't exist to pass to super constructor
    this.app = this;
    /**
    * Reference to the DependencyLoader instance
    * @type {DependencyLoader}
    */
    this.dependencyloader = new DependencyLoader(this);
    /**
    * Module for handling system configuration storage
    * @type {Object}
    */
    this.config = {};
    /**
    * Module for translating strings
    * @type {Object}
    */
    this.lang = {};
    /**
    * Module for logging messages
    * @type {Object}
    */
    this.logger = {};

    this.start();
  }
  /**
  * The module dependencies
  * @return {Object}
  */
  get dependencies() {
    return this.dependencyloader.dependencies;
  }
  /** @override */
  getConfig(key) {
    return this.app.config.get(`adapt-authoring-core.${key}`);
  }
  /**
  * Starts the app
  */
  async start() {
    const logError = e => {
      this.log(e.type, `${this.lang.t(e.key, e.data)}`);
      if(e.data.error) this.log('debug', e.data.error);
    };
    this.dependencyloader.on('loaderror', e => {
      logError(e);
      if(e.data.errors) e.data.errors.forEach(e => logError(e));
    });
    await this.dependencyloader.load();
    this.setReady();
    this.log('success', this.lang.t('info.startapp', { dir: this.getConfig('root_dir') }));
  }
  /**
  * Returns a module instance
  * @param {String} modName
  * @return {AbstractModule} The module instance
  */
  getModule(modName) {
    return this.dependencyloader.getModuleInstance(modName);
  }
  /**
  * Enables waiting for other modules to load
  * @param {String} modName
  * @return {Promise} Resolves when specified module has been loaded
  */
  async waitFor(modName) {
    return (await this.dependencyloader.waitFor(modName));
  }
}

module.exports = App;
